internal/exec/validate.go

```diff
@@ -0,0 +1,108 @@
+package exec
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/graph-gophers/graphql-go/errors"
+	"github.com/graph-gophers/graphql-go/internal/exec/resolvable"
+	"github.com/graph-gophers/graphql-go/internal/exec/selected"
+	"github.com/graph-gophers/graphql-go/types"
+)
+
+func collectFieldsToValidate(sels []selected.Selection, s *resolvable.Schema, fields *[]*fieldToValidate, fieldByAlias map[string]*fieldToValidate) {
+	for _, sel := range sels {
+		switch sel := sel.(type) {
+		case *selected.SchemaField:
+			field, ok := fieldByAlias[sel.Alias]
+			if !ok { // validation already checked for conflict (TODO)
+				field = &fieldToValidate{field: sel}
+				fieldByAlias[sel.Alias] = field
+				*fields = append(*fields, field)
+			}
+			field.sels = append(field.sels, sel.Sels...)
+		case *selected.TypenameField:
+			// TODO: confirm behaviour for this case
+			// Ignore __typename, which has no directives
+		case *selected.TypeAssertion:
+			collectFieldsToValidate(sel.Sels, s, fields, fieldByAlias)
+		default:
+			panic(fmt.Sprintf("unexpected selection type %T", sel))
+		}
+	}
+}
+
+func validateFieldSelection(ctx context.Context, s *resolvable.Schema, f *fieldToValidate, path *pathSegment) []*errors.QueryError {
+	if f.field.FixedResult.IsValid() {
+		// Skip fixed result meta fields like __TypeName
+		return nil
+	}
+
+	var args interface{}
+
+	if f.field.ArgsPacker != nil {
+		args = f.field.PackedArgs.Interface()
+	}
+
+	vErrs := f.field.Validate(ctx, args)
+
+	if l := len(vErrs); l > 0 {
+		errs := make([]*errors.QueryError, l)
+
+		for i, err := range vErrs {
+			var ext map[string]interface{}
+			if e, ok := err.(extensionser); ok {
+				ext = e.Extensions()
+			}
+			errs[i] = &errors.QueryError{
+				Err:        err,
+				Message:    err.Error(),
+				Locations:  []errors.Location{f.field.Loc}, // TODO: is this the correct location?
+				Path:       path.toSlice(),
+				Extensions: ext,
+			}
+		}
+
+		return errs
+	}
+
+	return validateSelectionSet(ctx, f.sels, f.field.Type, path, s)
+}
+
+func validateSelectionSet(ctx context.Context, sels []selected.Selection, typ types.Type, path *pathSegment, s *resolvable.Schema) []*errors.QueryError {
+	t, _ := unwrapNonNull(typ)
+
+	switch t.(type) {
+	case *types.ObjectTypeDefinition, *types.InterfaceTypeDefinition, *types.Union:
+		return validateSelections(ctx, sels, path, s)
+	}
+
+	switch t := t.(type) {
+	case *types.List:
+		return validateList(ctx, sels, t, path, s)
+	case *types.ScalarTypeDefinition:
+		// TODO: nothing needed here?
+	case *types.EnumTypeDefinition:
+		// TODO: nothing needed here?
```

Comment:
If this is the type of the field, then we still need to validate it, right? Although, I don't believe it is possible to have `EnumTypeDefinition` here. More like `EnumValue` if I am not mistaken. Easiest way to validate this is to add a unit tests where a field is of type custom scalar and another one where the result is of type enum. I am a bit confused about this switch case.

Reply:
Much of the validateX functions here are copied from the equivalent execX function (and adjusted to validate). These types are all handled there.

The types being dealt with here are the _schema_ types for the query selection I believe?